src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,211 @@ internal class DiagnosticComputer
+
+                // Step 2:
+                //  - Create the core 'computeTask' for computing diagnostics.
+                //  - Create a custom 'cancellationTokenSource' associated with this 'computeTask'.
+                //    This token source allows normal priority computeTasks to be cancelled when
+                //    a subsequent high priority diagnostic request is received.
+                using var cancellationTokenSource = new CancellationTokenSource();
+                var computeTask = CreateComputeTask(cancellationTokenSource, cancellationToken);
+
+                // Step 3:
+                //  - Add 'cancellationTokenSource' to the list of cancellation token sources for
+                //    currently executing normal priority tasks.
+                //    This list is used in 'GetHighPriorityDiagnosticsAsync' method to cancel all the
+                //    currently executing normal priority tasks if a new high priority diagnostic
+                //    request is subsequently received.
+                lock (s_gate)
+                {
+                    Debug.Assert(!s_cancellationSourcesForNormalPriorityComputeTasks.Contains(cancellationTokenSource));
+                    s_cancellationSourcesForNormalPriorityComputeTasks.Add(cancellationTokenSource);
+                }
+
+                try
+                {
+                    // Step 4:
+                    //  - Execute the core 'computeTask' for diagnostic computation.
+                    return await computeTask.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Step 5:
+                    // Check if cancellation fired on the custom 'cancellationTokenSource' that was created for
+                    // allowing cancellation of 'computeTask' from subsequent highPriority requests.
+                    if (cancellationTokenSource.IsCancellationRequested)
+                    {
+                        // Attempt to re-execute this cancelled normal priority task
+                        // by running the loop again.
+                        continue;
+                    }
+
+                    // Propagate all other OperationCanceledExceptions up the stack.
+                    throw;
+                }
+                finally
+                {
+                    // Step 6:
+                    //  - Remove the 'cancellationTokenSource' for completed or cancelled task.
+                    //    For the case where the computeTask was cancelled, we will create a new
+                    //    'cancellationTokenSource' for the retry.
+                    lock (s_gate)
+                    {
+                        var removed = s_cancellationSourcesForNormalPriorityComputeTasks.Remove(cancellationTokenSource);
+                        Debug.Assert(removed);
+                    }
+                }
+            }
+
+            throw ExceptionUtilities.Unreachable();
```

Comment:
i'm surprised this line is considered reachable by the compiler... there's no breaking out of hte while loop.  potential compile bug?

Reply:
I was able to remove this throw statement and code compiles fine. But, yes possibly compiler should have warned this statement as unreachable.